metadata_layer.py

```diff
@@ -437,6 +589,192 @@ def update(self, license_pool, license_pool_is_new):
         )
         return changed
 
+    @property
+    def has_open_access_link(self):
+        """Does this Circulation object have an associated open-access link?"""
+        return any(
+            [x for x in self.links 
+             if x.rel == Hyperlink.OPEN_ACCESS_DOWNLOAD and x.href]
+        )
+
+
+    def set_default_rights_uri(self, data_source, default_rights_uri=None):
+        """ TODO: default_rights_uri will be slightly different than old rights_uri, 
+        so decide on logic and then fix method.
+        """
+        if default_rights_uri == None and data_source:
+            # We didn't get rights passed in, so use the default rights for the data source if any.
+            default = RightsStatus.DATA_SOURCE_DEFAULT_RIGHTS_STATUS.get(data_source.name, None)
+            if default:
+                self.default_rights_uri = default
+
+        for format in self.formats:
+            if format.link:
+                link = format.link
+                if self.default_rights_uri in (None, RightsStatus.UNKNOWN) and link.rel == Hyperlink.OPEN_ACCESS_DOWNLOAD:
+                    # We haven't determined rights in the constructor or the data source, but there's an
+                    # open access download link, so we'll consider it generic open access.
+                    self.default_rights_uri = RightsStatus.GENERIC_OPEN_ACCESS
+
+        if self.default_rights_uri == None:
+            # We still haven't determined rights, so it's unknown.
+            self.default_rights_uri = RightsStatus.UNKNOWN
+
+
+    def apply(self, pool, replace=None):
+        """  TODO: apply name provisional, doc string follows.
+        """
+        made_changes = False
+
+        if pool is None:
+            raise ValueError("CirculationData.apply needs a pool.")
+
+        if replace is None:
+            replace = ReplacementPolicy()
+
+        _db = Session.object_session(pool)
+        data_source = self.data_source(_db)
+
+        identifier = pool.identifier
+
+        link_objects = {}
+        for link in self.links:
+            if link.rel in [Hyperlink.OPEN_ACCESS_DOWNLOAD, Hyperlink.DRM_ENCRYPTED_DOWNLOAD]:
+                link_obj, ignore = identifier.add_link(
+                    rel=link.rel, href=link.href, data_source=data_source, 
+                    license_pool=pool, media_type=link.media_type,
+                    content=link.content
+                )
+                link_objects[link] = link_obj
+
+        if pool and replace.formats:
+            for lpdm in pool.delivery_mechanisms:
+                _db.delete(lpdm)
+            pool.delivery_mechanisms = []
+
+        # TODO: bring back
+        #self.set_default_rights_uri(data_source)
+
+        for link in self.links:
+            if link.rel in [Hyperlink.OPEN_ACCESS_DOWNLOAD, Hyperlink.DRM_ENCRYPTED_DOWNLOAD]:
+                link_obj = link_objects[link]
+                # TODO: We do not properly handle the (unlikely) case
```

Comment:
I actually uncommented the set_default_rights_uri call, so it gets reset before the links start getting calculated.